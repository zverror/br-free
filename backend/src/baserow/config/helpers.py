import asyncio

from loguru import logger


class dummy_context:
    async def __aenter__(self):
        pass

    async def __aexit__(self, exc_type, exc, traceback):
        pass


class ConcurrencyLimiterASGI:
    """
    Helper wrapper on ASGI app to limit the number of requests handled
    at the same time.
    """

    def __init__(self, app, max_concurrency: int | None = None):
        self.app = app
        logger.info(f"Setting ASGI app concurrency to {max_concurrency}")
        self.semaphore = (
            asyncio.Semaphore(max_concurrency)
            if (isinstance(max_concurrency, int) and max_concurrency > 0)
            else dummy_context()
        )

    async def __call__(self, scope, receive, send):
        async with self.semaphore:
            await self.app(scope, receive, send)
