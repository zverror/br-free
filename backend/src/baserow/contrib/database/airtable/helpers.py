from typing import List

from baserow.contrib.database.fields.models import Field, SelectOption

from .constants import AIRTABLE_BASEROW_COLOR_MAPPING


def import_airtable_date_type_options(type_options) -> dict:
    date_format_mapping = {"European": "EU", "US": "US"}
    time_format_mapping = {"12hour": "12", "24hour": "24"}
    return {
        "date_format": date_format_mapping.get(type_options.get("dateFormat"), "ISO"),
        "date_include_time": type_options.get("isDateTime", False),
        "date_time_format": time_format_mapping.get(
            type_options.get("timeFormat"), "24"
        ),
    }


def import_airtable_choices(field_id: str, type_options: dict) -> List[SelectOption]:
    order = type_options.get("choiceOrder", [])
    choices = type_options.get("choices", [])
    return [
        SelectOption(
            # Combine select id with choice id as choice id is not guaranteed to be
            # unique across table
            id=f"{field_id}_{choice['id']}",
            value=choice["name"],
            color=AIRTABLE_BASEROW_COLOR_MAPPING.get(
                # The color isn't always provided, hence the fallback to an empty
                # color, which will fallback on the blue color.
                choice.get("color", ""),
                "blue",
            ),
            order=order.index(choice["id"]),
        )
        for choice in choices.values()
    ]


def set_select_options_on_field(
    field: Field, field_id: str, type_options: dict
) -> Field:
    """
    Set the `select_options` of a field in the prefetched objects cache. Overriding
    this object makes sure that when later `field.select_options.all()` is executed,
    it will return Airtable choices. This for example happens in
    `FieldType::export_serialized`.

    :param field: The field where the select must be set on.
    :param field_id: id of airtable choice field
    :param type_options: The options where to extract the Airtable choices from.
    :return: The updated field object.
    """

    # Overriding this object makes sure that when later
    # `field.select_options.all()` is executed, it will return Airtable choices.
    # This for example happens in `FieldType::export_serialized`.
    field._prefetched_objects_cache = {
        "select_options": import_airtable_choices(field_id, type_options)
    }
    return field
