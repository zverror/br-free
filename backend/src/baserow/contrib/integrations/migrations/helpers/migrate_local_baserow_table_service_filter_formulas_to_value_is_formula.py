"""
Helpers for 0006_migrate_local_baserow_table_service_filter_formulas_to_value_is_formula
"""

from baserow.core.formula import BaserowFormulaSyntaxError, get_parse_tree_for_formula


def value_parses_as_formula(value: str) -> bool:
    """
    Returns whether the service filter value is a formula or not.
    """

    # If the value is empty, it's not a formula.
    if not value:
        return False

    # If it's a string that's an integer, it's not a formula.
    try:
        int(value)
        return False
    except ValueError:
        pass

    try:
        get_parse_tree_for_formula(value)
        return True
    except BaserowFormulaSyntaxError:
        return False
