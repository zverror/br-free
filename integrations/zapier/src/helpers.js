const { unsupportedBaserowFieldTypes } = require('./constants')

/**
 * Fetches the fields of a table and converts them to an array with valid Zapier
 * field objects.
 */
const getRowInputValues = async (z, bundle) => {
  if (!bundle.inputData.tableID) {
    throw new Error('The `tableID` must be provided.')
  }

  const fieldsGetRequest = await z.request({
    url: `${bundle.authData.apiURL}/api/database/fields/table/${bundle.inputData.tableID}/`,
    method: 'GET',
    headers: {
      'Accept': 'application/json',
      'Authorization': `Token ${bundle.authData.apiToken}`,
    },
  })

  return fieldsGetRequest.json.map(v => {
    return mapBaserowFieldTypesToZapierTypes(v)
  })
}

/**
 * Fetches the fields and converts the input data to Baserow row compatible data.
 */
const prepareInputDataForBaserow = async (z, bundle) => {
  if (!bundle.inputData.tableID) {
    throw new Error('The `tableID` must be provided.')
  }

  const fieldsGetRequest = await z.request({
    url: `${bundle.authData.apiURL}/api/database/fields/table/${bundle.inputData.tableID}/`,
    method: 'GET',
    headers: {
      'Accept': 'application/json',
      'Authorization': `Token ${bundle.authData.apiToken}`,
    },
  })

  let rowData = { id: bundle.inputData.rowID }
  fieldsGetRequest
    .json
    .filter(
      (baserowField) =>
        baserowField.read_only
          || !unsupportedBaserowFieldTypes.includes(baserowField.type)
    )
    .filter((baserowField) => bundle.inputData.hasOwnProperty(baserowField.name))
    .forEach(baserowField => {
      let value = bundle.inputData[baserowField.name]

      if (baserowField.type === 'multiple_collaborators') {
        value = value.map(id => {
          return { id }}
        )
      }

      rowData[baserowField.name] = value
    })

  return rowData
}

/**
 * Converts the provided Baserow field type object to a Zapier compatible object.
 */
const mapBaserowFieldTypesToZapierTypes = (baserowField) => {
  const zapType = {
    key: baserowField.name,
    label: baserowField.name,
    type: 'string'
  }

  if (baserowField.type === 'long_text') {
    zapType.type = 'text'
  }

  if (baserowField.type === 'boolean') {
    zapType.type = 'boolean'
  }

  if (baserowField.type === 'number') {
    zapType.type = 'integer'

    if (baserowField.number_decimal_places > 0) {
      zapType.type = 'float'
    }
  }

  if (baserowField.type === 'boolean') {
    zapType.type = 'boolean'
  }

  if (baserowField.type === 'rating') {
    zapType.type = 'integer'
  }

  if (['single_select', 'multiple_select'].includes(baserowField.type)) {
    const choices = {}
    baserowField.select_options.forEach(el => {
      choices[`${el.id}`] = el.value
    })
    zapType.type = 'string'
    zapType.choices = choices
  }

  if (baserowField.type === 'multiple_select') {
    zapType.list = true
  }

  if (baserowField.type === 'link_row') {
    zapType.type = 'integer'
    zapType.helpText = 'Provide row ids that you want to link to.'
    zapType.list = true
  }

  if (baserowField.type === 'multiple_collaborators') {
    zapType.type = 'integer'
    zapType.helpText = 'Provide user ids that you want to link to.'
    zapType.list = true
  }

  if (baserowField.type === 'date' && !baserowField.date_include_time) {
    zapType.type = 'date'
    zapType.helpText =
      'the date fields accepts a date in ISO format (e.g. 2020-01-01)'
  }

  if (baserowField.type === 'date' && baserowField.date_include_time) {
    zapType.type = 'datetime'
    zapType.helpText =
      'the date fields accepts date and time in ISO format (e.g. 2020-01-01 12:00)'
  }

  if (
    baserowField.read_only
    || unsupportedBaserowFieldTypes.includes(baserowField.type)
  ) {
    // Read only and the file field are not supported.
    return
  }

  return zapType
}

module.exports = {
  getRowInputValues,
  prepareInputDataForBaserow,
  mapBaserowFieldTypesToZapierTypes,
}
