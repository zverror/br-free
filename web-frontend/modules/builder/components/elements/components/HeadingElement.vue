<template>
  <div
    :class="{
      'element--no-value': !resolvedValue,
    }"
  >
    <ABHeading :level="element.level" :style="getStyleOverride('typography')">
      {{
        element.value
          ? resolvedValue ||
            (mode === 'editing' ? $t('headingElement.emptyValue') : '&nbsp;')
          : $t('headingElement.missingValue')
      }}
    </ABHeading>
  </div>
</template>

<script>
import element from '@baserow/modules/builder/mixins/element'
import { ensureString } from '@baserow/modules/core/utils/validator'

export default {
  name: 'HeadingElement',
  mixins: [element],
  props: {
    /**
     * @type {Object}
     * @property {number} level - The size of the heading
     * @property {string} value - The text displayed
     */
    element: {
      type: Object,
      required: true,
    },
  },
  computed: {
    resolvedValue() {
      return ensureString(this.resolveFormula(this.element.value))
    },
  },
}
</script>
