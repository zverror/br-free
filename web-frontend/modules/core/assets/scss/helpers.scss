%clearfix {
  &::after {
    clear: both;
    content: '';
    display: table;
  }
}

.clearfix {
  @extend %clearfix;
}

.hidden {
  display: none;
}

.visibility-hidden {
  visibility: hidden;
  transform: translateX(-10000px) translateY(-10000px);
}

.align-right {
  text-align: right;
}

.background-white {
  background-color: $white;
}

.color-primary {
  color: $color-primary-500 !important;
}

.color-success {
  color: $color-success-500 !important;
}

.color-warning {
  color: $color-warning-500 !important;
}

.color-error {
  color: $color-error-500 !important;
}

.color-neutral {
  color: $color-neutral-500 !important;
}

.padding-top-2 {
  padding-top: 16px !important;
}

.margin-top-0 {
  margin-top: 0 !important;
}

.margin-top-1 {
  margin-top: 8px !important;
}

.margin-top-2 {
  margin-top: 16px !important;
}

.margin-top-3 {
  margin-top: 30px !important;
}

.margin-top-4 {
  margin-top: 40px !important;
}

.margin-bottom-0 {
  margin-bottom: 0 !important;
}

.margin-bottom-1 {
  margin-bottom: 8px !important;
}

.margin-bottom-2 {
  margin-bottom: 16px !important;
}

.margin-bottom-24 {
  margin-top: 24px !important;
}

.margin-bottom-3 {
  margin-bottom: 30px !important;
}

.margin-bottom-4 {
  margin-bottom: 40px !important;
}

.margin-left-auto {
  margin-left: auto !important;
}

.margin-left-0 {
  margin-left: 0 !important;
}

.margin-left-1 {
  margin-left: 8px !important;
}

.margin-left-2 {
  margin-left: 16px !important;
}

.margin-left-3 {
  margin-left: 30px !important;
}

.margin-left-4 {
  margin-left: 40px !important;
}

.margin-right-auto {
  margin-right: auto !important;
}

.margin-right-0 {
  margin-right: 0 !important;
}

.margin-right-1 {
  margin-right: 8px !important;
}

.margin-right-2 {
  margin-right: 16px !important;
}

.margin-right-3 {
  margin-right: 30px !important;
}

.margin-right-4 {
  margin-right: 40px !important;
}

.resizing-horizontal {
  cursor: col-resize;
}

// This selector can only be used to temporarily show an element for calculating
// positions.
.forced-block {
  display: block !important;
}

// Sometimes we forcefully want to disable the user-select property to initial
// because the element is for example being edited via a contenteditable attribute
// which does not work in combination with the user-select: none; property in Safari.
.forced-user-select-initial {
  user-select: initial !important;
}

.forced-text-overflow-initial {
  text-overflow: initial !important;
}

.forced-pointer-events-auto {
  pointer-events: auto !important;
}

.prevent-scroll {
  overflow: hidden !important;
}

.whitespace-pre-wrap {
  white-space: pre-wrap;
}

.remove-number-input-controls,
%remove-number-input-controls {
  appearance: textfield !important;

  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    appearance: none;
    margin: 0;
  }
}

.flex {
  width: 100%;
  display: flex;
  gap: var(--gap, 10px);
}

.flex-wrap {
  flex-wrap: wrap;
}

.flex-100 {
  width: 100%;
  flex-basis: 100%;
}

.flex-grow-1 {
  flex-grow: 1;
}

.align-items-center {
  align-items: center;
}

.justify-content-space-between {
  justify-content: space-between;
}

.position-relative {
  position: relative;
}

.inline-block {
  display: inline-block;
}

@keyframes spin {
  0% {
    transform: rotate(0);
  }

  50% {
    transform: rotate(360deg);
  }

  100% {
    transform: rotate(720deg);
  }
}

/* spacer classes generation */

@each $size in $spacings {
  .p-#{$size},
  %p-#{$size} {
    padding: #{$size}px;
  }

  .px-#{$size},
  %px-#{$size} {
    padding-left: #{$size}px;
    padding-right: #{$size}px;
  }

  .py-#{$size},
  %py-#{$size} {
    padding-top: #{$size}px;
    padding-bottom: #{$size}px;
  }

  .m-#{$size},
  %m-#{$size} {
    margin: #{$size}px;
  }

  .my-#{$size},
  %my-#{$size} {
    margin-top: #{$size}px;
    margin-bottom: #{$size}px;
  }

  .mx-#{$size},
  %mx-#{$size} {
    margin-left: #{$size}px;
    margin-right: #{$size}px;
  }

  .pl-#{$size},
  %pl-#{$size} {
    padding-left: #{$size}px;
  }

  .ml-#{$size},
  %ml-#{$size} {
    margin-left: #{$size}px;
  }

  .pr-#{$size},
  %pr-#{$size} {
    padding-right: #{$size}px;
  }

  .mr-#{$size},
  %mr-#{$size} {
    margin-right: #{$size}px;
  }

  .pb-#{$size},
  %pb-#{$size} {
    padding-bottom: #{$size}px;
  }

  .mb-#{$size},
  %mb-#{$size} {
    margin-bottom: #{$size}px;
  }

  .pt-#{$size},
  %pt-#{$size} {
    padding-top: #{$size}px;
  }

  .mt-#{$size},
  %mt-#{$size} {
    margin-top: #{$size}px;
  }
}
