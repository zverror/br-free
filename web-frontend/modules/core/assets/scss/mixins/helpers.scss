@mixin fixed-height($height, $font-size) {
  height: $height;
  line-height: $height;
  font-size: $font-size;
}

@mixin center-text($width, $font-size, $height: $width) {
  @include fixed-height($height, $font-size);

  width: $width;
  text-align: center;
}

@mixin flex-align-items($gap: 0, $display: flex) {
  display: $display;
  align-items: center;
  gap: $gap;
}

@mixin loading($size: 1.4rem) {
  position: relative;
  display: block;
  width: $size;
  height: $size;
  border-radius: 50%;
  border: 0.25em solid;
  border-color: $color-primary-500 transparent $color-primary-500 transparent;
  animation: spin infinite 1800ms;
  animation-timing-function: cubic-bezier(0.785, 0.135, 0.15, 0.86);
}

// The basically works the same as padding: 10px 0 0 10px or padding: 10px; only then
// resulting position: absolute; with corresponding top, right, bottom, left values.
// -------
// Example:
//  @include absolute(10px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 10px; bottom: 10px; left: 10px;
// -------
// Example:
//  @include absolute(10px, 20px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 10px; left: 20px;
// -------
// Example:
//  @include absolute(10px, 20px, 30px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 30px;
@mixin absolute($top: null, $right: null, $bottom: null, $left: null) {
  position: absolute;

  @if $top != null and $right == null and $bottom == null and $left == null {
    inset: $top $top $top $top;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else if $top != null and $right != null and $bottom == null and $left == null
  {
    inset: $top $right $top $right;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else {
    @if $top {
      top: $top;
    }

    @if $right {
      right: $right;
    }

    @if $bottom {
      bottom: $bottom;
    }

    @if $left {
      left: $left;
    }
  }
}
