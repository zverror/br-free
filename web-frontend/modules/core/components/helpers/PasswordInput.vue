<!--
This helper component can only be used in a form which uses the password validation
from modules/core/validators.js
as ist expects the following vuelidate validation state params

- required
- maxLength
- minLength

In order to use this component in another component/form it needs a value
prop which can be passed via the 'v-model' directive. That means the state for the 
password value always lives in the parent component.
The parent component also needs to provide the vuelidate password validation in order
for this component to show all the possible validation errors.

Example Usage:

<PasswordInput
   v-model="PASSWORDSTATE"
   :validation-state="the vuelidate validation instance of the given password value"
/>

In case the password validation rules change, the passwordValidation object in the
validators file needs to be updated, as well as this component.

modules/core/validators.js

-->
<template>
  <FormGroup
    small-label
    :label="label"
    required
    :error="validationState.$error"
  >
    <FormInput
      :error="validationState.$error"
      type="password"
      size="large"
      :autocomplete="autocomplete"
      :value="value"
      :placeholder="placeholder"
      @blur="validationState.$touch()"
      @input="$emit('input', $event)"
    >
    </FormInput>

    <template #error>
      <span v-if="validationState.$error && !validationState.required">
        <i v-if="showErrorIcon" class="iconoir-warning-triangle"></i>
        {{ $t('error.inputRequired') }}
      </span>
      <span v-if="validationState.$error && !validationState.maxLength">
        <i v-if="showErrorIcon" class="iconoir-warning-triangle"></i>
        {{
          $t('error.maxLength', {
            max: validationState.$params.maxLength.max,
          })
        }}
      </span>
      <span v-if="validationState.$error && !validationState.minLength">
        <i v-if="showErrorIcon" class="iconoir-warning-triangle"></i>
        {{
          $t('error.minLength', {
            min: validationState.$params.minLength.min,
          })
        }}
      </span>
    </template>
  </FormGroup>
</template>

<script>
export default {
  name: 'PasswordInput',
  props: {
    validationState: {
      type: Object,
      required: true,
    },
    value: {
      type: String,
      required: true,
    },
    label: {
      type: String,
      required: false,
      default: null,
    },
    autocomplete: {
      type: String,
      required: false,
      default: 'new-password',
    },
    placeholder: {
      type: String,
      required: false,
      default: null,
    },
    errorPlaceholderClass: {
      type: String,
      required: false,
      default: '',
    },
    showErrorIcon: {
      type: Boolean,
      required: false,
      default: false,
    },
  },
}
</script>
