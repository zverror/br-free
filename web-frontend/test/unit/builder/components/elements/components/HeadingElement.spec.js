import { TestApp } from '@baserow/test/helpers/testApp'
import HeadingElement from '@baserow/modules/builder/components/elements/components/HeadingElement.vue'

describe('HeadingElement', () => {
  let testApp = null
  let store = null

  beforeAll(() => {
    testApp = new TestApp()
    store = testApp.store
    store.$registry.registerNamespace('element')
  })

  afterEach(() => {
    testApp.afterEach()
  })

  const mountComponent = ({ props = {}, slots = {}, provide = {} }) => {
    return testApp.mount(HeadingElement, {
      propsData: props,
      slots,
      provide,
    })
  }

  test('Default HeadingElement component', async () => {
    const builder = { id: 1, theme: { primary_color: '#ccc' } }
    const page = {}
    const workspace = {}
    const element = { level: 2, value: '', styles: {} }
    const mode = 'public'

    const wrapper = await mountComponent({
      props: {
        element,
      },
      provide: {
        builder,
        mode,
        currentPage: page,
        elementPage: page,
        applicationContext: { builder, page, mode },
        element,
        workspace,
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  test('Default HeadingElement component v2', async () => {
    const builder = { id: 1, theme: { primary_color: '#ccc' } }
    const page = {}
    const workspace = {}
    const element = { level: 3, value: '"hello"', styles: {} }
    const mode = 'public'

    const wrapper = await mountComponent({
      props: {
        element,
      },
      provide: {
        builder,
        mode,
        currentPage: page,
        elementPage: page,
        applicationContext: { builder, page, mode },
        element,
        workspace,
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
