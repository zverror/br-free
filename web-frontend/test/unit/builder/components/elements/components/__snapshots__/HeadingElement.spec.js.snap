// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeadingElement Default HeadingElement component 1`] = `
<div
  class="element--no-value"
>
  <h2
    class="ab-heading ab-heading--h2"
  >
    
    headingElement.missingValue
  
  </h2>
</div>
`;

exports[`HeadingElement Default HeadingElement component v2 1`] = `
<div
  class=""
>
  <h3
    class="ab-heading ab-heading--h3"
  >
    
    hello
  
  </h3>
</div>
`;
